libflux/flux-core/src/semantic/types.rs

```diff
@@ -1256,10 +1256,14 @@ impl Substitutable for Dynamic {
 
 impl Dynamic {
     fn constrain(&self, with: Kind, _: &mut Substitution) -> Result<(), Error> {
-        Err(Error::CannotConstrain {
-            act: MonoType::Dynamic(self.clone().into()),
-            exp: with,
-        })
+        match with {
+            // Dynamic needs to be able to go into the various cast functions
+            Kind::Basic => Ok(()),
+            _ => Err(Error::CannotConstrain {
+                act: MonoType::Dynamic(self.clone().into()),
+                exp: with,
+            }),
+        }
```

Comment:
This prevents `dynamic` being passed into functions that specify `Record` (or anything else). `Basic` was needed to continue to allow the standard cast fns to work, although I didn't actually find where they specify this. Still, the tests failed complaining the "arg must be `Basic`".

iirc when we were discussing the design, we raised the possibility of needing a new Kind for `Basic + Dynamic`. Not sure it's necessary to add a new Kind at this stage, but I just now remembered it was mentioned.